src/Data/Nat/Primality.agda

```diff
@@ -8,80 +8,290 @@
+composite⇒nonTrivial {1}      = flip contradiction ¬composite[1]
+composite⇒nonTrivial {2+ _} _ = _
+
+prime⇒nonTrivial : Prime p → NonTrivial p
+prime⇒nonTrivial (prime ⦃ nt ⦄ _) = nt
+
+------------------------------------------------------------------------
+-- NonZero
+
+composite⇒nonZero : Composite n → NonZero n
+composite⇒nonZero {suc _} _ = _
+
+prime⇒nonZero : Prime p → NonZero p
+prime⇒nonZero (prime _) = nonTrivial⇒nonZero
+
+irreducible⇒nonZero : Irreducible n → NonZero n
+irreducible⇒nonZero {zero} = flip contradiction ¬irreducible[0]
+irreducible⇒nonZero {suc _} _ = _
 
-Prime : ℕ → Set
-Prime 0 = ⊥
-Prime 1 = ⊥
-Prime n = ∀ {d} → 2 ≤ d → d < n → d ∤ n
 
 ------------------------------------------------------------------------
 -- Decidability
 
 composite? : Decidable Composite
-composite? 0               = no λ()
-composite? 1               = no λ()
-composite? n@(suc (suc _)) = Dec.map′
-  (map₂ λ { (a , b , c) → (b , a , c)})
-  (map₂ λ { (a , b , c) → (b , a , c)})
-  (anyUpTo? (λ d → 2 ≤? d ×-dec d ∣? n) n)
+composite? n = Dec.map′
+  (map₂ λ (d<n , 1<d , d∣n) → boundedComposite>1 1<d d<n d∣n)
+  (map₂ λ (boundedComposite d<n d∣n) → d<n , nonTrivial⇒n>1 , d∣n)
+  (anyUpTo? (λ d → 1 <? d ×-dec d ∣? n) n)
 
 prime? : Decidable Prime
-prime? 0               = no λ()
-prime? 1               = no λ()
-prime? n@(suc (suc _)) = Dec.map′
-  (λ f {d} → flip (f {d}))
-  (λ f {d} → flip (f {d}))
-  (allUpTo? (λ d → 2 ≤? d →-dec ¬? (d ∣? n)) n)
+prime? 0       = no ¬prime[0]
+prime? 1       = no ¬prime[1]
+prime? n@(2+ _) = Dec.map′
+  (λ r → prime λ (boundedComposite d<n d∣n) → r d<n nonTrivial⇒n>1 d∣n)
+  (λ (prime p) {d} d<n 1<d d∣n → p {d} (boundedComposite>1 1<d d<n d∣n))
+  (allUpTo? (λ d → 1 <? d →-dec ¬? (d ∣? n)) n)
+
+irreducible? : Decidable Irreducible
+irreducible? zero      = no ¬irreducible[0]
+irreducible? n@(suc _) = Dec.map′ bounded-irr⇒irr irr⇒bounded-irr
+  (allUpTo? (λ m → (m ∣? n) →-dec ((m ≟ 1) ⊎-dec m ≟ n)) n)
+  where
+  BoundedIrreducible : Pred ℕ _
+  BoundedIrreducible n = ∀ {m} → m < n → m ∣ n → m ≡ 1 ⊎ m ≡ n
+  bounded-irr⇒irr : BoundedIrreducible n → Irreducible n
+  bounded-irr⇒irr bounded-irr m∣n
+    = [ flip bounded-irr m∣n , inj₂ ]′ (m≤n⇒m<n∨m≡n (∣⇒≤ m∣n))
+  irr⇒bounded-irr : Irreducible n → BoundedIrreducible n
+  irr⇒bounded-irr irr m<n m∣n = irr m∣n
 
 ------------------------------------------------------------------------
--- Relationships between compositeness and primality
+-- Relationships between compositeness, primality, and irreducibility
 
 composite⇒¬prime : Composite n → ¬ Prime n
-composite⇒¬prime {n@(suc (suc _))} (d , 2≤d , d<n , d∣n) n-prime =
-  n-prime 2≤d d<n d∣n
+composite⇒¬prime (d , composite[d]) (prime p) = p composite[d]
 
-¬composite⇒prime : 2 ≤ n → ¬ Composite n → Prime n
-¬composite⇒prime (s≤s (s≤s _)) ¬n-composite {d} 2≤d d<n d∣n =
-  ¬n-composite (d , 2≤d , d<n , d∣n)
+¬composite⇒prime : ⦃ NonTrivial n ⦄ → ¬ Composite n → Prime n
```

Comment:
First argument should be irrelevant?

Reply:
So the coda to all these comments is that, having moved `NonTrivial` to `Data.Nat.Base`, a lot of explicit instances need to be put back in... already suggesting that the `{{nt}}` fields of `BoundedComposite` and `Prime` need to be made irrelevant, and then... that might be the way to tackle your comments in the whole?

Reply:
DONE!